From c1612c906cd42c2766e802fc3a6756b91a0cb712 Mon Sep 17 00:00:00 2001
From: Dhanavandhana Kannan <quic_dhanavan@quicinc.com>
Date: Mon, 29 Apr 2024 09:50:00 +0530
Subject: [PATCH] hostapd: Add support to get the link mac address in Beacon
 request frame

Currently for Beacon Request frame, station's MLD address
will be sent via hostapd_cli command and that address will
be used to send the frame to the station.

Suppose if station's link address is given, it results in the
failure of the command as AP was not able to find the
station with the station's link address.

The fix is to provide support to get station link address
which is mentioned in hostapd_cli command for Beacon request
frame.

Signed-off-by: Dhanavandhana Kannan <quic_dhanavan@quicinc.com>
---
 src/ap/rrm.c | 14 +++++++++-----
 1 file changed, 9 insertions(+), 5 deletions(-)

--- a/src/ap/rrm.c
+++ b/src/ap/rrm.c
@@ -644,11 +644,15 @@ int hostapd_send_beacon_req(struct hosta
 	pos = wpabuf_head(req);
 	mode = pos[6];
 
-	if (!sta || !(sta->flags & WLAN_STA_AUTHORIZED)) {
-		wpa_printf(MSG_INFO,
-			   "Beacon request: " MACSTR " is not connected",
-			   MAC2STR(addr));
-		return -1;
+	if (!sta) {
+		if (hapd->mld)
+			sta = ap_get_link_sta(hapd, addr);
+		if (!sta || !(sta->flags & WLAN_STA_AUTHORIZED)) {
+			wpa_printf(MSG_INFO,
+				   "Beacon request: " MACSTR " is not connected",
+				   MAC2STR(addr));
+			return -1;
+		}
 	}
 
 	switch (mode) {
@@ -709,7 +713,7 @@ int hostapd_send_beacon_req(struct hosta
 	wpabuf_put_u8(buf, MEASURE_TYPE_BEACON); /* Measurement Type */
 	wpabuf_put_buf(buf, req);
 
-	ret = hostapd_drv_send_action(hapd, hapd->iface->freq, 0, addr,
+	ret = hostapd_drv_send_action(hapd, hapd->iface->freq, 0, sta->addr,
 				      wpabuf_head(buf), wpabuf_len(buf));
 	wpabuf_free(buf);
 	if (ret < 0)
